[DXIL] Define and generate `DXILAttribute` and `DXILProperty`

inbelic opened the pull request:
- Redefines `DXILAttribute` to denote a function attribute, compatible to how it was define in DXC/LLVM 3.7
- Fix how `DXILAttribute` is emitted to be a struct of set attributes instead of an "or" of the enums
- Implement the lowering of `DXILAttribute` to LLVM function attributes in `DXILOpBuilder.cpp`. A custom mapping is defined.
- Audit all current ops to specify the correct attributes consistent with DXC. This is done here to allow for testing.
- Update testcases in `llvm/test/CodeGen/DirectX` of all ops with attributes to match that attributes are set
- Update testcases of ops that had previously incorrectly set attributes to check there is no attributes set
- Defines `DXILProperty` to denote the other type of attributes from DXC used to query properties.
- Emit `DXILProperty` as a struct of set attributes.
- Updates `DXIL.td` to specify applicable `DXILProperty`s on ops

Note: `DXILProperty` was referred to as 'queryable attributes' in design discussion. Changed to property to allow for better expression in `DXIL.td`

Resolves #114461
Resolves #115912

inbelic committed:
[DXIL] Define DXILAttribute

- switch to using DXILAttribute to only denote function attributes

- attribute enums can't be or'd together as is currently implemented, so
  we switch to using a list of attributes in DXILOpBuilder.cpp and
  DXILEmitter.cpp

inbelic committed:
correct ReadNone/ReadOnly function attributes

- correct all uses of ReadOnly/ReadNone to be consistent with hctdb.py
in DXC
- all fix order of attributes in each op

inbelic committed:
update testcases to check attributes

- testcases are update to check updated attribute types
- there are some CHECK-NOT tests to ensure that previously set
attributes no longer emit an attribute

inbelic committed:
define DXILProperty for queryable attributes

- define these properties in DXIL.td and DXILConstants.h
- emit DXIL definitions as enumerations
- emit some helper functions to query OpCodeProp for each property class

inbelic commented:
In design discussion I used the terms 'function attribute' vs 'queryable attribute'. In this commit I change the terminology to `DXILAttribute` for 'function attribute' and `DXILProperty` for 'queryable attribute'.

Reviewer suggestions are very welcomed to this naming.

inbelic committed:
clang format

inbelic committed:
change to emitting a general query helper

- instead of generating a query helper for each property, just use a
general one for all property enums

inbelic committed:
add newly added op on rebase

inbelic commented on llvm/utils/TableGen/DXILEmitter.cpp:
```diff
@@ -396,19 +396,16 @@ static void emitDXILProperties(const RecordKeeper &Records, raw_ostream &OS) {
-    OS << "  return false;\n";
-    OS << "}\n\n";
-  }
+static void emitDXILPropertyHelper(raw_ostream &OS) {
```
This won't be tested until we use it in https://github.com/llvm/llvm-project/issues/114565. So we could wait until there to implement this

pow2clk commented on llvm/lib/Target/DirectX/DXIL.td:
```diff
@@ -266,18 +266,30 @@ def miss : DXILShaderStage;
+class DXILProperty;
+
+def IsBarrier : DXILProperty;
+def IsDerivative : DXILProperty;
```
This was set, but never actually used in hctdb.py. What was actually used is the is_gradient value. This change demonstrates it can be removed from DXC with no consequences: https://github.com/microsoft/DirectXShaderCompiler/pull/7023

As such, I think we should remove this property

inbelic replied:
Furthermore, do we want to undefine all the properties and only define them when an op that uses them is added? Currently all but `isWave`. It would prevent having any other redundant properties by being easier to track.

inbelic committed:
remove dependcy on OpCodeProperty for helper function

- before this change, we would be required to define OpCodeProperty and
all used structs to use hasProperty, which for all uses outside of
DXILOpBuilder is not usable.

inbelic committed:
review comment:

- remove `IsDerivative` property as it was not used in `hctdb.py`

inbelic commented:
Rebased to resolve existing conflicts. I also added two more commits to this pr regarding formatting which will let us also resolve https://github.com/llvm/llvm-project/issues/115912. In hindsight, this pr already completed most of the work tasks for both issues, so it seems better to just have one pr rather than going back and detangling the tasks into 2 prs.

inbelic committed:
update `DXILProperty` uses

- update existing wave ops to specify the `IsWave` property
- update the `Barrier` op to specify the `IsBarrier` property

llvm-beanz commented on llvm/lib/Target/DirectX/DXIL.td:
```diff
@@ -51,195 +51,194 @@ def SplitDoubleTy : DXILOpParamType;
 
 class DXILOpClass;
 
-defset list<DXILOpClass> OpClasses = {
```
The formatting changes make this diff a lot harder to read and much larger than it really should be. In general if you have a large format-only diff it is best to put that in a separate NFC PR so that the functional change can be reviewed separately.

Can you break this out into a separate formatting-only NFC change?

inbelic committed:
Merge branch 'main' into inbelic/dxil-gen

inbelic committed:
update for recently added intrinsic

bogner commented:
Personally I think we should rip out all of the codegen for properties until we're actually doing something with them. It might be fine to have them in DXIL.td if we want to copy these over as we go, but the DXILEmitter stuff really shouldn't be there until we're using the properties in code.

Note that this whole approach, for both attributes and properties, is continuing a bit of an antipattern that's already there. It would be much nicer if we used tablegen to generate tables rather than using is as a generic c++ code emitter.

Consider, in DXILConstants.h we could define a structure instead of an enum:
```c++
struct Properties {
#define DXIL_PROPERTY(Name) bool Name;
#include "DXILOperation.inc"
};
```

Then, in DXILOpBuilder rather than adding to the OpCodeProperty mess we could simpy have a function like:
```c++
static void setDXILProperties(dxil::OpCode OpCode, CallInst *CI) {
  dxil::Properties Props;
  switch (OpCode) {
#define DXIL_OP_PROPERTIES(OpCode, ...)                                        \
  case OpCode:                                                                 \
    Props = {__VA_ARGS__};                                                     \
    break;
#include "DXILOperation.inc"
  }
  if (Props.IsBarrier)
    ...
  if (Props.IsGradient)
    ...
  if (Props.IsFeedback)
    ...
  if (Props.IsWave)
    ...
  if (Props.RequiresUniformInputs)
    ...
}
```

Where instead of building a string of all of the properties we simply emit a table of bools:
```
// op xyz is a barrier and a gradient, but not feedback, wave, or requires uniform inputs
DXIL_OP_PROPERTIES(dxil::opcode::xyz, true, true, false, false, false)
```

A similar approach should probably be applied to attributes, but it's a bit more complicated if we want to have the versioning so I think it's reasonable to do that as part of a more general cleanup of DXILEmitter later.

inbelic committed:
fix testcase typo

inbelic committed:
review comments: remove attribute string generation

- convert to using a struct to store the active attributes to let us use
a table of bools instead of a string list of attributes

inbelic committed:
review comments: remove properties codegen

- as noted, the codegen was not going to be used until a future pr
- it is now removed and we restructure the properties to be a struct of
bools, with the intention to remove any codegen and follow a similar
pattern to the attributes in the future

inbelic committed:
clang formatting

llvm-beanz commented on llvm/utils/TableGen/DXILEmitter.cpp:
```diff
@@ -177,6 +178,13 @@ DXILOperationDesc::DXILOperationDesc(const Record *R) {
+  Recs = R->getValueAsListOfDefs("properties");
+
+  // Get property records
+  for (const Record *CR : Recs) {
+    PropRecs.push_back(CR);
+  }
```
nit:
```suggestion
  for (const Record *CR : Recs)
    PropRecs.push_back(CR);
```
see: https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

llvm-beanz reviewed (approved):
LGTM, one really small formatting nit.

pow2clk reviewed (approved):
I wrote a few different comments before I resolved them myself. Following the way the versioned op attributes are matched was a bit confusing, but I'm convinced it's right now. What remains are attributes that match what we had in hctdb, but I wonder if they should be different. We can certainly consider that later.

inbelic committed:
[NFC] review comments:

- update nit from review comments
- fix some typos in comments
- add additional comments for clarity

inbelic commented on llvm/lib/Target/DirectX/DXIL.td:
```diff
@@ -783,6 +797,7 @@ def CreateHandle : DXILOp<57, createHandle> {
   let arguments = [Int8Ty, Int32Ty, Int32Ty, Int1Ty];
   let result = HandleTy;
   let stages = [Stages<DXIL1_0, [all_stages]>, Stages<DXIL1_6, [removed]>];
+  let attributes = [Attributes<DXIL1_0, [ReadOnly]>];
```
Added a comment to note this and will mention it in the new issue.

inbelic commented on llvm/lib/Target/DirectX/DXIL.td:
```diff
@@ -974,7 +996,7 @@ def WaveAllBitCount : DXILOp<135, waveAllOp> {
   let result = Int32Ty;
   let stages = [Stages<DXIL1_0, [all_stages]>];
-  let attributes = [Attributes<DXIL1_0, [ReadNone]>];
+  let properties = [IsWave];
```
Yes, I agree that we could further investigate the correctness of the attributes. The goal of this PR was to get consistent behaviour with DXC, so I will file an issue to look into that.

inbelic commented:
Created an issue to track the additional work of verifying the current attributes specified in DXC: https://github.com/llvm/llvm-project/issues/123825
